### PR TITLE
[IMP] stock: wave pickings

### PR DESCRIPTION
**Pull request by Whenrow**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Whenrow** [comment]
@all sorry for the unwanted review request.

File: odoo/addons/base/models/ir_model.py [hunk 1]
```diff
@@ -1897,7 +1897,7 @@ def xmlid_to_res_id(self, xmlid, raise_if_not_found=False):
 
     @api.model
     def xmlid_to_object(self, xmlid, raise_if_not_found=False):
-        """ Return a Model object, or ``None`` if ``raise_if_not_found`` is 
+        """ Return a Model object, or ``None`` if ``raise_if_not_found`` is
```
Comment (odony) [inline odoo/addons/base/models/ir_model.py:1900]: someone has an issue with their linter autofixing files? ;-)

File: addons/stock_picking_wave/models/stock_picking_wave.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        view = self.env.ref('stock_picking_wave.view_immediate_wave')
+        return {
+            'name': _('Immediate Transfer?'),
+            'type': 'ir.actions.act_window',
+            'view_mode': 'form',
+            'res_model': 'stock.immediate.wave',
+            'views': [(view.id, 'form')],
+            'view_id': view.id,
+            'target': 'new',
+            'context': dict(self.env.context, default_show_waves=show_waves, default_wave_ids=[(4, w.id) for w in self]),
```
Comment (xmo-odoo) [inline addons/stock_picking_wave/models/stock_picking_wave.py:228]: Commands!

File: addons/stock_picking_wave/wizard/stock_immediate_wave.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                (0, 0, {'to_immediate': True, 'wave_id': wave_id})
+                for wave_id in res['wave_ids'][0][2]
+            ]
+        return res
+
+    def process(self):
+        waves_to_do = self.env['stock.picking.wave']
+        waves_not_to_do = self.env['stock.picking.wave']
+        for line in self.immediate_wave_line_ids:
+            if line.to_immediate is True:
```
Comment (xmo-odoo) [inline addons/stock_picking_wave/wizard/stock_immediate_wave.py:41]:
```suggestion
            if line.to_immediate:
```
?

File: addons/stock_picking_wave/wizard/stock_immediate_wave.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        'stock.immediate.wave.line',
+        'immediate_wave_id',
+        string="Immediate Wave Lines")
+
+    @api.model
+    def default_get(self, fields_list):
+        res = super().default_get(fields_list)
+        if 'immediate_wave_line_ids' in fields_list and res.get('wave_ids'):
+            res['immediate_wave_line_ids'] = [
+                (0, 0, {'to_immediate': True, 'wave_id': wave_id})
```
Comment (xmo-odoo) [inline addons/stock_picking_wave/wizard/stock_immediate_wave.py:32]: Commands :badger:

**Comment by Whenrow** [comment]
@xmo-odoo Thanks for the review even if it was due to a prankster rebase and a non-checked lazy push :angel:

File: addons/stock_picking_wave/models/stock_picking_wave.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+        company = self.env.company
+        picking_type = self.env['stock.picking.type'].search([
+            ('company_id', '=', company.id),
+            ('code', '=', 'incoming')
+        ], limit=1)
+        return picking_type or False
+
+    name = fields.Char(
+        string='Wave Transfer', default='New',
+        copy=False, required=True, readonly=True,
```
[the remark this answers is not in the document]
Comment (Whenrow) [reply addons/stock_picking_wave/models/stock_picking_wave.py:24]: because of the sequence. Like the picking name

File: addons/stock_picking_wave/models/stock_picking_wave.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+    user_id = fields.Many2one(
+        'res.users', string='Responsible', tracking=True, check_company=True,
+        readonly=True, states={'draft': [('readonly', False)], 'in_progress': [('readonly', False)]},
+        help='Person responsible for this wave transfer')
+    company_id = fields.Many2one(
+        'res.company', string="Company", required=True, readonly=True,
+        index=True, default=lambda self: self.env.company)
+    picking_ids = fields.Many2many(
+        'stock.picking', help='List of transfers associated to this wave')
+    allowed_picking_ids = fields.One2many('stock.picking', compute='_compute_allowed_picking_ids')
```
[the remark this answers is not in the document]
Comment (Whenrow) [reply addons/stock_picking_wave/models/stock_picking_wave.py:35]: I dislike @ryv-odoo

File: addons/web/static/src/legacy/js/fields/relational_fields.js [hunk 7]
```diff
@@ -2004,7 +2004,7 @@ var FieldOne2Many = FieldX2Many.extend({
 
     /**
      * @private
-     * @param {*} data 
+     * @param {*} data
```
Comment (SimonGenin) [inline addons/web/static/src/legacy/js/fields/relational_fields.js:2007]: unnecessary diff

File: addons/web/static/src/legacy/js/fields/relational_fields.js [hunk 8]
```diff
@@ -1349,7 +1349,7 @@ var FieldX2Many = AbstractField.extend(WidgetAdapterMixin, {
             if (extraInfo.subFieldName) {
                 parts.push(`[name="${extraInfo.subFieldName}"]`);
             }
-    
+
```
Comment (SimonGenin) [inline addons/web/static/src/legacy/js/fields/relational_fields.js:1352]: unnecessary diff

**Review by SimonGenin — CHANGES_REQUESTED** [review]

File: addons/stock_picking_wave/models/stock_picking_wave.py [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+            move_line_ids = self.move_line_ids.filtered(
+                lambda ml:
+                float_compare(ml.qty_done, 0.0, precision_rounding=ml.product_uom_id.rounding) > 0
+                and not ml.result_package_id
+            )
+            if not move_line_ids:
+                move_line_ids = self.move_line_ids.filtered(
+                    lambda ml:
+                    float_compare(ml.product_uom_qty, 0.0, precision_rounding=ml.product_uom_id.rounding) > 0
+                    and float_compare(ml.qty_done, 0.0, precision_rounding=ml.product_uom_id.rounding) == 0
```
Comment (amoyaux) [inline addons/stock_picking_wave/models/stock_picking_wave.py:183]: and not ml.result_package_id also no?

File: addons/stock_picking_wave/models/stock_picking_wave.py [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+            raise UserError(_("Please add 'Done' quantities to the wave picking to create a new pack."))
+
+    # -------------------------------------------------------------------------
+    # Miscellaneous
+    # -------------------------------------------------------------------------
+    def _pre_put_in_pack_hook(self, move_line_ids):
+        res = self.move_line_ids.picking_id[0]._check_destinations(move_line_ids)
+        if res and res['res_id']:
+            self.env['stock.package.destination'].browse(res['res_id']).wave_id = self
+        return res
```
Comment (amoyaux) [inline addons/stock_picking_wave/models/stock_picking_wave.py:199]: you could override _check_destination instead an add 'wave_id' in the returned action

File: addons/stock_picking_wave/models/stock_picking_wave.py [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+                    "Please check their states and operation types, if they aren't immediate "
+                    "transfers or if they're not already part of another wave transfer.\n\n"
+                    "Incompatibilities: %s", wave.name, ', '.join(erroneous_pickings.mapped('name'))))
+
+    def _track_subtype(self, init_values):
+        if 'state' in init_values:
+            return self.env.ref('stock_picking_wave.mt_wave_state')
+        return super()._track_subtype(init_values)
+
+    def _should_show_waves(self):
```
Comment (amoyaux) [inline addons/stock_picking_wave/models/stock_picking_wave.py:245]: not sure we need this method

File: addons/stock_picking_wave/static/src/js/choose_lines_widget.js [hunk 12]
```diff
@@ -0,0 +1,67 @@
+odoo.define('stock_picking_wave.choose_lines_widget', function (require) {
```
Comment (amoyaux) [inline addons/stock_picking_wave/static/src/js/choose_lines_widget.js:1]: Check if it's possible to already do it in owl

File: addons/stock_picking_wave/models/stock_picking_wave.py [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+            if move.move_line_ids <= self.move_line_ids:
+                continue
+            qty = sum([line.qty_done for line in move.move_line_ids & self.move_line_ids])
+            qty_split = move.product_uom._compute_quantity(move.product_uom_qty - qty, move.product_id.uom_id, rounding_method='HALF-UP')
+            new_move = move._split(qty_split, waves=self)
+            new_move_vals += new_move
+        if new_move_vals:
+            new_moves = self.env['stock.move'].create(new_move_vals)
+            new_moves._action_confirm(merge=False)
+        self.move_line_ids.move_id._action_done()
```
Comment (amoyaux) [inline addons/stock_picking_wave/models/stock_picking_wave.py:165]: Don't forget to:
trigger_assign on validate move

+ _action_done on stock.picking if last line was done in the wave

**Comment by amoyaux** [comment]
Globally modify the idea. A wave is a stock.picking, it will be easier to maintain as discussed

File: addons/stock/views/stock_picking_views.xml [hunk 14]
```diff
@@ -635,6 +635,28 @@
             </field>
         </record>
 
+        <record id="action_get_picking_type_operations" model="ir.actions.act_window">
+            <field name="name">Operations</field>
+            <field name="res_model">stock.move.line</field>
+            <field name="type">ir.actions.act_window</field>
+            <field name="view_mode">tree,form</field>
```
Comment (amoyaux) [inline addons/stock/views/stock_picking_views.xml:642]: not needed since view_ids is set I think. + it should be only tree no?

File: addons/stock_picking_batch/models/stock_move_line.py [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+class StockMoveLine(models.Model):
+    _inherit = "stock.move.line"
+
+    def action_add_to_wave(self):
+        wave = self.env.context.get('active_wave')
+        if not wave:
+            return
+        self._change_picking(wave)
+
+    def _change_picking(self, wave=False):
```
Comment (amoyaux) [inline addons/stock_picking_batch/models/stock_move_line.py:19]: wave as arg

File: addons/stock_picking_batch/models/stock_move_line.py [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+from odoo import models
+from odoo import Command
+
+
+class StockMoveLine(models.Model):
+    _inherit = "stock.move.line"
+
+    def action_add_to_wave(self):
+        wave = self.env.context.get('active_wave')
```
Comment (amoyaux) [inline addons/stock_picking_batch/models/stock_move_line.py:14]: wave_id

File: addons/stock_picking_batch/views/stock_move_line_views.xml [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+        <field name="inherit_id" ref="stock.stock_move_line_view_search"/>
+        <field name="arch" type="xml">
+            <xpath expr="//filter[@name='by_date']" position="after">
+                <filter string="Transfers" name="by_picking" domain="[]" context="{'group_by': 'picking_id'}"/>
+                <filter string="Location" name="by_location" domain="[]" context="{'group_by': 'location_id'}"/>
+            </xpath>
+        </field>
+    </record>
+
+    <record id="view_stock_move_operations_to_wave" model="ir.ui.view">
```
Comment (amoyaux) [inline addons/stock_picking_batch/views/stock_move_line_views.xml:15]: You can probably do a primary copy of the view in stock now

File: addons/stock_picking_batch/views/stock_move_line_views.xml [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+                <field name="package_id" groups="stock.group_tracking_lot"/>
+                <field name="result_package_id" groups="stock.group_tracking_lot" optional="hide"/>
+                <field name="product_uom_qty"/>
+                <field name="product_uom_id"/>
+                <field name="qty_done"/>
+            </tree>
+        </field>
+    </record>
+
+    <record id="view_move_line_tree" model="ir.ui.view">
```
Comment (amoyaux) [inline addons/stock_picking_batch/views/stock_move_line_views.xml:40]: check if still needed

File: addons/stock_picking_batch/wizard/stock_picking_to_wave.py [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+
+class StockMoveLineToWave(models.TransientModel):
+    _name = 'stock.move.line.to.wave'
+    _description = 'Wave Transfer Lines'
+
+    wave_id = fields.Many2one('stock.picking', string='Wave Transfer')
+    line_ids = fields.Many2many('stock.move.line')
+
+    @api.model
+    def default_get(self, fields_list):
```
Comment (amoyaux) [inline addons/stock_picking_batch/wizard/stock_picking_to_wave.py:64]: before fields declaration

File: addons/stock_picking_batch/wizard/stock_picking_to_wave.py [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+class StockPickingToWave(models.TransientModel):
+    _name = 'stock.picking.to.wave'
+    _description = 'Wave Transfer Lines'
+
+    wave_id = fields.Many2one('stock.picking.batch', string='Wave Transfer', domain="[('is_wave', '=', True), ('state', '!=', 'done')]")
+    mode = fields.Selection([('existing', 'an existing wave transfer'), ('new', 'a new wave transfer')], default='existing')
+    user_id = fields.Many2one('res.users', string='Responsible', help='Person responsible for this wave transfer')
+
+    @api.model
+    def default_get(self, fields_list):
```
Comment (amoyaux) [inline addons/stock_picking_batch/wizard/stock_picking_to_wave.py:17]: before fields declaration

File: addons/stock/models/stock_picking.py [hunk 21]
```diff
@@ -523,6 +523,27 @@ def _compute_date_deadline(self):
             else:
                 picking.date_deadline = max(picking.move_lines.filtered('date_deadline').mapped('date_deadline'), default=False)
 
+    def action_add_operations(self):
```
[the remark this answers is not in the document]
Comment (amoyaux) [reply addons/stock/models/stock_picking.py:541]: still need to be done

**Comment by amoyaux** [comment]
robodoo r+ rebase-ff

**Comment by Whenrow** [comment]
robodoo r-

**Comment by amoyaux** [comment]
robodoo r+

**Comment by amoyaux** [comment]
robodoo r+
